### ISSUE TITLE
Batch Playlist Lookup

### ISSUE THREAD

**Issue by emmx** [claudiob/yt#213]
Hello,

I've a number of playlists, for each playlist I want to process all its videos, and for each video I want to get some data (title, view_count, like_count, comment_count, ...).

The problem is that I'm using the following code, which doesn't fetch the statistics in advance so it takes about 1s for each video, which is way too long.

How do I download all the statistics in advance?

``` ruby
# Example

> playlist = Yt::Playlist.new(id: "PL55713C70BA91BD6E")
=> #<Yt::Models::Playlist:0x00000003fdb788 @auth=nil, @id="PL55713C70BA91BD6E">

> playlist.playlist_items.map {|item| puts item.video.view_count}
594934110
281465291
138438681
44146678
51594886
803499943
```

**Comment by claudiofullscreen** [comment]
There is an `.includes` method that should help you with that.

On Thu, Jul 2, 2015 at 5:51 PM, Matt notifications@github.com wrote:

> Hello,
> 
> I've a number of playlists, for each playlist I want to process all its
> videos, and for each video I want to get some data (title, view_count,
> like_count, comment_count, ...).
> 
> The problem is that I'm using the following code, which doesn't fetch the
> statistics in advance so it takes about 1s for each video, which is way too
> long.
> 
> How do I download all the statistics in advance?
> 
> # Example
> 
> > playlist = Yt::Playlist.new(id: "PL55713C70BA91BD6E")
> > => #<Yt::Models::Playlist:0x00000003fdb788 @auth=nil, @id="PL55713C70BA91BD6E">
> > playlist.playlist_items.map {|item| puts item.video.view_count}5949341102814652911384386814414667851594886803499943
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Fullscreen/yt/issues/213.

## 

 Claudio Baccigalupo  | Software Writer  |  Fullscreen, Inc.
http://Fullscreen.net  |  +1.310.202.3333

**Comment by emmx** [comment]
Can you please give me an example? I couldn't find how to fix the code I published above, it takes too long regardless of what I put in `.includes`. Thanks!

**Comment by emmx** [comment]
Any idea? .includes is not working with playlists.

**Comment by claudiofullscreen** [comment]
I double-checked and you're right, it's not there for now.

However, I'll need this feature soon, so I'll build it. 
The idea is for the following snippet of code to retrieve all snippets and statistics at once

``` ruby
playlist.playlist_items.includes(video: [:snippet, :statistics]).each do |item|
  puts "Views: #{item.video.view_count}"
end
```

**Comment by emmx** [comment]
That would be awesome. I'll wait for that push, thanks!

**Comment by emmx** [comment]
Any news on this issue?

**Comment by emmx** [comment]
Hey guys, I really need this feature. Do you plan to fix this soon? I guess I can fork it and send a pull request but I barely know how this gem works. Cheers!

**Comment by stavro** [comment]
@mparodi 

Use this in the meantime:

``` ruby
playlist.playlist_items.each_slice(50) do |playlist_item_batch|
  yt_videos = Yt::Collections::Videos.new.where(id: playlist_item_batch.map(&:video_id).join(','), part: 'snippet,statistics')
  yt_videos.each do |yt_video|
    # You may now access in bulk:
    #  yt_video.title
    #  yt_video.id
    #  yt_video.view_count
    #  yt_video.comment_count
    #  yt_video.like_count
    #  yt_video.dislike_count
    #  yt_video.favorite_count
    #  etc...
  end
end
```

**Comment by emmx** [comment]
Thanks, it's an improvement but it still takes +20 seconds to process only 250 videos.

**Comment by claudiofullscreen** [comment]
@mparodi Rather than implementing the feature I described above as:

``` ruby
playlist.playlist_items.includes(video: [:snippet, :statistics]).each do |item|
  puts "Views: #{item.video.view_count}"
end
```

I decided to start by simply adding `.includes(:video)` to `playlist_items`, which will eager-load (for each video) the snippet, status and statistics.

This should solve your issue. For instance, following your example, this code:

``` ruby
playlist = Yt::Playlist.new(id: "PL55713C70BA91BD6E")
playlist.playlist_items.includes(:video).map {|item| item.video.view_count rescue "Private"}
```

used to take 85 seconds, but now only takes 14 seconds if you update to `0.25.2`.

_(Note that the `rescue "Private"` is there because there is a private video in the list you pointed to)_
